rhel/vex/updater.go

```diff
@@ -0,0 +1,185 @@
+package vex
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/quay/zlog"
+
+	"github.com/quay/claircore/libvuln/driver"
+)
+
+var (
+	_ driver.Updater           = (*Updater)(nil)
+	_ driver.Configurable      = (*Updater)(nil)
+	_ driver.DeltaUpdater      = (*Updater)(nil)
+	_ driver.UpdaterSetFactory = (*Factory)(nil)
+	_ driver.Configurable      = (*Factory)(nil)
+)
+
+const (
+	// BaseURL is the base url for the Red Hat VEX security data.
+	//
+	//doc:url updater
+	BaseURL = "https://access.redhat.com/security/data/csaf/v2/vex/"
+
+	latestFile     = "archive_latest.txt"
+	changesFile    = "changes.csv"
+	deletionsFile  = "deletions.csv"
+	lookBackToYear = 2014
+	repoKey        = "rhel-cpe-repository"
+	updaterVersion = "1"
+)
+
+// Factory creates an Updater to process all of the Red Hat VEX data.
+//
+// [Configure] must be called before [UpdaterSet].
+type Factory struct {
+	c    *http.Client
+	base *url.URL
+}
+
+// UpdaterSet constructs one Updater
+func (f *Factory) UpdaterSet(_ context.Context) (driver.UpdaterSet, error) {
+	us := driver.NewUpdaterSet()
+	u := &Updater{
+		url:    f.base,
+		client: f.c,
+	}
+	err := us.Add(u)
+	if err != nil {
+		return us, err
+	}
+	return us, nil
+}
+
+// FactoryConfig is the configuration honored by the Factory.
+//
+// The URL is where the updater expects the VEX data to be published
+// (and must end with a slash).
+type FactoryConfig struct {
+	// URL indicates the base URL for the VEX.
+	//
+	// Must include the trailing slash.
+	URL string `json:"url" yaml:"url"`
+}
+
+// Configure implements driver.Configurable
+func (f *Factory) Configure(ctx context.Context, cf driver.ConfigUnmarshaler, c *http.Client) error {
+	f.c = c
+	var cfg FactoryConfig
+	if err := cf(&cfg); err != nil {
+		return err
+	}
+	var err error
+	u := BaseURL
+	if cfg.URL != "" {
+		u = cfg.URL
+		if !strings.HasSuffix(u, "/") {
+			u += "/"
+		}
+	}
+	f.base, err = url.Parse(u)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Updater is responsible from reading VEX data served at the URL
+// and creating vulnerabilities.
+type Updater struct {
+	url    *url.URL
+	client *http.Client
+}
+
+// fingerprint is used to track the state of the changes.csv and deletions.csv endpoints.
+//
+// The spec (https://www.rfc-editor.org/rfc/rfc9110.html#name-etag) mentions
+// that there is no need for the client to be aware of how each entity tag
+// is constructed, however, it mentions that servers should avoid backslashes.
```

Comment:
I added the comment in case it wasn't obvious why backslashes were used, haven't seen then used at any point in the etags coming from RH security data servers.